### PR TITLE
Simulated CPU benchmarking code 

### PR DESCRIPTION
**Pull request by Yash9060**
This PR adds simulated benchmarking code for CPU, this version can be run directly on cloudtop as well as can be integrated into Kokoro runs (Hence it unblocks [b/361427435](https://b.corp.google.com/361427435) 

Also sharing code between demo & benchmarks so that there is no discrepancy between both the versions.  

- [X] Tests pass

<details>
<summary>Logs from sample run</summary>
```
(.venv) yashsha@yashsha1:~/vscode/dataflux-pytorch$ python3 -u demo/lightning/checkpoint/simulated/multiprocessing_train.py --project=gcs-tess, --ckpt-dir-path=gs://yashsha-us-east1-d/ --model-size=300 --world-size=5
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:490: UserWarning: Detected an existing checkpoint in gs:/yashsha-us-east1-d/checkpoints/ckpt_0.ckpt/.metadata, overwriting since self.overwrite=True. Past version 2.5 of PyTorch, `overwrite` will default to False. Set this variable to True to maintain this functionality or False to raise when an existing checkpoint is found.
  warnings.warn(
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:490: UserWarning: Detected an existing checkpoint in gs:/yashsha-us-east1-d/checkpoints/ckpt_0.ckpt/.metadata, overwriting since self.overwrite=True. Past version 2.5 of PyTorch, `overwrite` will default to False. Set this variable to True to maintain this functionality or False to raise when an existing checkpoint is found.
  warnings.warn(
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:490: UserWarning: Detected an existing checkpoint in gs:/yashsha-us-east1-d/checkpoints/ckpt_0.ckpt/.metadata, overwriting since self.overwrite=True. Past version 2.5 of PyTorch, `overwrite` will default to False. Set this variable to True to maintain this functionality or False to raise when an existing checkpoint is found.
  warnings.warn(
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:490: UserWarning: Detected an existing checkpoint in gs:/yashsha-us-east1-d/checkpoints/ckpt_0.ckpt/.metadata, overwriting since self.overwrite=True. Past version 2.5 of PyTorch, `overwrite` will default to False. Set this variable to True to maintain this functionality or False to raise when an existing checkpoint is found.
  warnings.warn(
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:490: UserWarning: Detected an existing checkpoint in gs:/yashsha-us-east1-d/checkpoints/ckpt_0.ckpt/.metadata, overwriting since self.overwrite=True. Past version 2.5 of PyTorch, `overwrite` will default to False. Set this variable to True to maintain this functionality or False to raise when an existing checkpoint is found.
  warnings.warn(
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:116: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  if tensor.storage().size() != tensor.numel():
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:116: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  if tensor.storage().size() != tensor.numel():
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:116: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  if tensor.storage().size() != tensor.numel():
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:116: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  if tensor.storage().size() != tensor.numel():
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:116: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  if tensor.storage().size() != tensor.numel():
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:490: UserWarning: Detected an existing checkpoint in gs:/yashsha-us-east1-d/checkpoints/ckpt_1.ckpt/.metadata, overwriting since self.overwrite=True. Past version 2.5 of PyTorch, `overwrite` will default to False. Set this variable to True to maintain this functionality or False to raise when an existing checkpoint is found.
  warnings.warn(
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:490: UserWarning: Detected an existing checkpoint in gs:/yashsha-us-east1-d/checkpoints/ckpt_1.ckpt/.metadata, overwriting since self.overwrite=True. Past version 2.5 of PyTorch, `overwrite` will default to False. Set this variable to True to maintain this functionality or False to raise when an existing checkpoint is found.
  warnings.warn(
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:490: UserWarning: Detected an existing checkpoint in gs:/yashsha-us-east1-d/checkpoints/ckpt_1.ckpt/.metadata, overwriting since self.overwrite=True. Past version 2.5 of PyTorch, `overwrite` will default to False. Set this variable to True to maintain this functionality or False to raise when an existing checkpoint is found.
  warnings.warn(
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:490: UserWarning: Detected an existing checkpoint in gs:/yashsha-us-east1-d/checkpoints/ckpt_1.ckpt/.metadata, overwriting since self.overwrite=True. Past version 2.5 of PyTorch, `overwrite` will default to False. Set this variable to True to maintain this functionality or False to raise when an existing checkpoint is found.
  warnings.warn(
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:490: UserWarning: Detected an existing checkpoint in gs:/yashsha-us-east1-d/checkpoints/ckpt_1.ckpt/.metadata, overwriting since self.overwrite=True. Past version 2.5 of PyTorch, `overwrite` will default to False. Set this variable to True to maintain this functionality or False to raise when an existing checkpoint is found.
  warnings.warn(
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:490: UserWarning: Detected an existing checkpoint in gs:/yashsha-us-east1-d/checkpoints/ckpt_2.ckpt/.metadata, overwriting since self.overwrite=True. Past version 2.5 of PyTorch, `overwrite` will default to False. Set this variable to True to maintain this functionality or False to raise when an existing checkpoint is found.
  warnings.warn(
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:490: UserWarning: Detected an existing checkpoint in gs:/yashsha-us-east1-d/checkpoints/ckpt_2.ckpt/.metadata, overwriting since self.overwrite=True. Past version 2.5 of PyTorch, `overwrite` will default to False. Set this variable to True to maintain this functionality or False to raise when an existing checkpoint is found.
  warnings.warn(
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:490: UserWarning: Detected an existing checkpoint in gs:/yashsha-us-east1-d/checkpoints/ckpt_2.ckpt/.metadata, overwriting since self.overwrite=True. Past version 2.5 of PyTorch, `overwrite` will default to False. Set this variable to True to maintain this functionality or False to raise when an existing checkpoint is found.
  warnings.warn(
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:490: UserWarning: Detected an existing checkpoint in gs:/yashsha-us-east1-d/checkpoints/ckpt_2.ckpt/.metadata, overwriting since self.overwrite=True. Past version 2.5 of PyTorch, `overwrite` will default to False. Set this variable to True to maintain this functionality or False to raise when an existing checkpoint is found.
  warnings.warn(
/usr/local/google/home/yashsha/vscode/.venv/lib/python3.11/site-packages/torch/distributed/checkpoint/filesystem.py:490: UserWarning: Detected an existing checkpoint in gs:/yashsha-us-east1-d/checkpoints/ckpt_2.ckpt/.metadata, overwriting since self.overwrite=True. Past version 2.5 of PyTorch, `overwrite` will default to False. Set this variable to True to maintain this functionality or False to raise when an existing checkpoint is found.
  warnings.warn(
######################
Time taken to save checkpoint: 2.2351 seconds
Time taken to load checkpoint: 0.2796 seconds
Size of distributed tensors (rank 0): 68.80 MB
Total size of all tensors (rank 0): 344.01 MB
######################
```
</details>

### CODE REVIEW

File: dataflux_pytorch/benchmark/checkpointing/simulated/multi_node_benchmark.py [hunk 1]
```diff
@@ -2,52 +2,17 @@
...
+    SimpleModel,
+    cleanup,
+    format_size,
+    get_tensor_size_bytes,
+    split_tensor,
+    time_checkpoint_operation,
+)
 
 
 def configure_master_addr():
```
Comment (jdnurme) [inline dataflux_pytorch/benchmark/checkpointing/simulated/multi_node_benchmark.py:18]: does this function need to exist if we always expect to run localhost as our host? Is there an example where we don't do that?

File: demo/lightning/checkpoint/simulated/multiprocessing_train.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        description="Benchmark distributed checkpointing.")
+    parser.add_argument("--project", type=str,
+                        required=True, help="GCS project ID.")
+    parser.add_argument("--ckpt-dir-path", type=str, required=True,
+                        help="Path to GCS bucket for checkpoints.")
+    parser.add_argument("--model-size", type=int,
+                        default=100, help="Size of the model.")
+    parser.add_argument("--clear-kernel-cache", action="store_true",
+                        default=False, help="Clear kernel cache.")
+    parser.add_argument("--sample-size", type=int, default=3,
```
Comment (jdnurme) [inline demo/lightning/checkpoint/simulated/multiprocessing_train.py:75]: Should this be renamed --sample-count if its the number of iterations?

File: demo/lightning/checkpoint/simulated/multiprocessing_train.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    return tensor.view(-1)[start_idx:end_idx]
+
+
+def time_checkpoint_operation(benchmark_strategy: BenchmarkStrategy, distributed_state_dict: Dict[str, torch.Tensor], filepath: str, sample_size: int, operation: str) -> list:
+    """Times save or load operations for checkpoints."""
+    times = []
+    for i in range(sample_size):
+        start_time = time.time()
+        if operation == 'save':
+            checkpoint_path = os.path.join(
```
Comment (jdnurme) [inline demo/lightning/checkpoint/simulated/multiprocessing_train.py:122]: checkpoint_path could be moved outside of the if/elif blocks.

**Review by jdnurme — APPROVED** [review]

File: demo/lightning/checkpoint/simulated/multiprocessing_train.py (hunk 2 above)
Comment (Yash9060) [reply demo/lightning/checkpoint/simulated/multiprocessing_train.py:75]: Makes sense. Done

File: demo/lightning/checkpoint/simulated/multiprocessing_train.py (hunk 3 above)
Comment (Yash9060) [reply demo/lightning/checkpoint/simulated/multiprocessing_train.py:122]: Done

File: dataflux_pytorch/benchmark/checkpointing/simulated/multi_node_benchmark.py (hunk 1 above)
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/multi_node_benchmark.py:18]: We do need this in GKE setup, since application runs on multiple pods across different nodes. Each node will have its own IP address.

File: demo/lightning/checkpoint/simulated/multiprocessing_train.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+
+class BenchmarkStrategy(FSDPStrategy):
+    def __init__(self, project: str, path: str, model: nn.Module, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.writer = GCSDistributedWriter(path, project, None)
+        self.reader = GCSDistributedReader(path, project, None)
+        self.model = model
+
+    def save_checkpoint(self, checkpoint: Dict[str, torch.Tensor], filepath: str, storage_options: Optional[Dict] = None) -> None:
```
Comment (MattIrv) [inline demo/lightning/checkpoint/simulated/multiprocessing_train.py:41]: It may help to add some doc comments throughout this PR, especially explaining `save_checkpoint` and `load_checkpoint` here

File: demo/lightning/checkpoint/simulated/multiprocessing_train.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+def parse_args() -> argparse.Namespace:
+    """Parses command line arguments."""
+    parser = argparse.ArgumentParser(
+        description="Benchmark distributed checkpointing.")
+    parser.add_argument("--project", type=str,
+                        required=True, help="GCS project ID.")
+    parser.add_argument("--ckpt-dir-path", type=str, required=True,
+                        help="Path to GCS bucket for checkpoints.")
+    parser.add_argument("--model-size", type=int,
+                        default=100, help="Size of the model.")
```
Comment (MattIrv) [inline demo/lightning/checkpoint/simulated/multiprocessing_train.py:72]: What unit is this in? Is it the number of layers?

File: demo/lightning/checkpoint/simulated/multiprocessing_train.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    parser.add_argument("--clear-kernel-cache", action="store_true",
+                        default=False, help="Clear kernel cache.")
+    parser.add_argument("--sample-count", type=int, default=3,
+                        help="Number of samples for benchmarking.")
+    parser.add_argument("--padding-size", type=int, default=1000,
+                        help="Size of dummy tensors for padding, to control the size of the checkpoint.")
```
Comment (MattIrv) [inline demo/lightning/checkpoint/simulated/multiprocessing_train.py:78]: Is there any way to add some info on how this setting affects the size? e.g. "1000 is 10 MB" or something like that?

File: demo/lightning/checkpoint/simulated/multiprocessing_train.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+            filepath, f'checkpoints/ckpt_{i}.ckpt')
+        start_time = time.time()
+        if operation == 'save':
+            benchmark_strategy.save_checkpoint(
+                distributed_state_dict, filepath=checkpoint_path)
+        elif operation == 'load':
+            benchmark_strategy.load_checkpoint(checkpoint_path=checkpoint_path)
+        end_time = time.time()
+        times.append(end_time - start_time)
+        dist.barrier()  # Synchronize processes
```
Comment (MattIrv) [inline demo/lightning/checkpoint/simulated/multiprocessing_train.py:130]: You might want to print the save/load time here from rank 0. When saves/loads take a long time, this will help the user have confidence that it is making progress.

File: demo/lightning/checkpoint/simulated/multiprocessing_train.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    return tensor.view(-1)[start_idx:end_idx]
+
+
+def time_checkpoint_operation(benchmark_strategy: BenchmarkStrategy, distributed_state_dict: Dict[str, torch.Tensor], filepath: str, sample_count: int, operation: str) -> list:
+    """Times save or load operations for checkpoints."""
+    times = []
+    for i in range(sample_count):
+        checkpoint_path = os.path.join(
+            filepath, f'checkpoints/ckpt_{i}.ckpt')
+        start_time = time.time()
```
Comment (MattIrv) [inline demo/lightning/checkpoint/simulated/multiprocessing_train.py:122]: Do we need a dist.barrier() before this?

**Review by MattIrv — APPROVED** [review]

File: demo/lightning/checkpoint/simulated/multiprocessing_train.py (hunk 5 above)
Comment (Yash9060) [reply demo/lightning/checkpoint/simulated/multiprocessing_train.py:72]: It is not number of layers. But size of layers (Number of layer is fixed to 2 in above example) I will change this to layer_size.

File: demo/lightning/checkpoint/simulated/multiprocessing_train.py (hunk 7 above)
Comment (Yash9060) [reply demo/lightning/checkpoint/simulated/multiprocessing_train.py:130]: Good point. Will do !

File: demo/lightning/checkpoint/simulated/multiprocessing_train.py (hunk 8 above)
Comment (Yash9060) [reply demo/lightning/checkpoint/simulated/multiprocessing_train.py:122]: Done

File: demo/lightning/checkpoint/simulated/multiprocessing_train.py (hunk 4 above)
Comment (Yash9060) [reply demo/lightning/checkpoint/simulated/multiprocessing_train.py:41]: Add some docstrings.

File: demo/lightning/checkpoint/simulated/multiprocessing_train.py (hunk 6 above)
Comment (Yash9060) [reply demo/lightning/checkpoint/simulated/multiprocessing_train.py]: Added. It approximately adds 3 MB per 100 units of padding-size.
